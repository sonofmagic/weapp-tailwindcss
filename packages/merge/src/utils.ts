import type { ClassValue } from 'clsx'
import type {
  createTailwindMerge as _createTailwindMergeV3,
  extendTailwindMerge as _extendTailwindMergeV3,
  twJoin as _twJoinV3,
  twMerge as _twMergeV3,
} from 'tailwind-merge'
import type {
  createTailwindMerge as _createTailwindMergeV2,
  extendTailwindMerge as _extendTailwindMergeV2,
  twJoin as _twJoinV2,
  twMerge as _twMergeV2,
} from 'tailwind-merge-v2'
import type { CreateOptions } from './types'
import { clsx } from 'clsx'
import { escape } from 'weapp-tailwindcss/escape'

export type TailwindMergeVersion = 2 | 3
// 处理 escape 逻辑
export function noop(x: string) {
  return x
}

export interface CreateFactoryOptions<V extends TailwindMergeVersion> {
  version: V
  twMerge: V extends 3 ? typeof _twMergeV3 : typeof _twMergeV2
  twJoin: V extends 3 ? typeof _twJoinV3 : typeof _twJoinV2
  extendTailwindMerge: V extends 3 ? typeof _extendTailwindMergeV3 : typeof _extendTailwindMergeV2
  createTailwindMerge: V extends 3 ? typeof _createTailwindMergeV3 : typeof _createTailwindMergeV2
}

export function createFactory<V extends TailwindMergeVersion>(opts: CreateFactoryOptions<V>) {
  const {
    twMerge: _twMerge,
    twJoin: _twJoin,
    createTailwindMerge: _createTailwindMerge,
    extendTailwindMerge: _extendTailwindMerge,
  } = opts

  return function create(options?: CreateOptions) {
    const escapeFn = options?.escapeFn ?? escape
    const e = options?.disableEscape ? noop : escapeFn

    function twMerge(...inputs: ClassValue[]): string {
      return e(_twMerge(clsx(...inputs)))
    }

    function twJoin(...inputs: ClassValue[]): string {
      return e(_twJoin(clsx(...inputs)))
    }

    // 版本安全的 extendTailwindMerge
    const extendTailwindMerge = (...args: Parameters<typeof _extendTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _extendTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    // 版本安全的 createTailwindMerge
    const createTailwindMerge = (...args: Parameters<typeof _createTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _createTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    return {
      twMerge,
      twJoin,
      extendTailwindMerge,
      createTailwindMerge,
    }
  }
}
