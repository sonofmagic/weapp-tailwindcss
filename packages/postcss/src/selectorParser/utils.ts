// 选择器 AST 相关的通用工具，提供组合与替换的辅助函数
import type { Node } from 'postcss-selector-parser'
import type { IStyleHandlerOptions } from '../types'
import psp from 'postcss-selector-parser'

export type ParserTransformOptions = Partial<{
  lossless: boolean
  updateSelector: boolean
}>

// normalizeTransformOptions 确保 parser 在修改选择器时保持必要的副作用
export function normalizeTransformOptions(options?: ParserTransformOptions): ParserTransformOptions {
  return {
    lossless: false,
    updateSelector: true,
    ...options,
  }
}

// mklist 将节点扩展为 "节点 + combinator + 节点副本" 的结构
export function mklist(node: Node): Node[] {
  return [
    node,
    psp.combinator({
      value: '+',
    }),
    node.clone(),
  ]
}

// composeIsPseudoAst 按需把字符串数组转换为 :is(...) 结构的 AST
export function composeIsPseudoAst(strs: string | string[]): Node[] {
  if (typeof strs === 'string') {
    return mklist(psp.tag({
      value: strs,
    }))
  }
  if (strs.length > 1) {
    return mklist(psp.pseudo({
      value: ':is',
      nodes: strs.map(str =>
        psp.tag({
          value: str,
        }),
      ),
    }))
  }
  return mklist(psp.tag({
    value: strs[0],
  }))
}

// 根据配置生成替换子代选择器的 AST，默认使用 view 标签
export function getCombinatorSelectorAst(options: IStyleHandlerOptions) {
  let childCombinatorReplaceValue: Node[] = mklist(psp.tag({ value: 'view' }))
  const { cssChildCombinatorReplaceValue } = options
  if (
    typeof cssChildCombinatorReplaceValue === 'string'
    || (Array.isArray(cssChildCombinatorReplaceValue) && cssChildCombinatorReplaceValue.length > 0)) {
    childCombinatorReplaceValue = composeIsPseudoAst(cssChildCombinatorReplaceValue)
  }
  return childCombinatorReplaceValue
}
