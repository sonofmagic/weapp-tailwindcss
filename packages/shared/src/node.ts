import crypto from 'node:crypto'
import path from 'pathe'

export function md5(input: crypto.BinaryLike) {
  return crypto.createHash('md5').update(input).digest('hex')
}

export function removeAllExtensions(filePath: string) {
  if (!filePath) {
    return ''
  }

  let baseName = path.basename(filePath)
  let ext = path.extname(baseName)

  while (ext) {
    baseName = baseName.slice(0, -ext.length)
    ext = path.extname(baseName)
  }

  const dir = path.dirname(filePath)
  return dir === '.' ? baseName : path.join(dir, baseName)
}
