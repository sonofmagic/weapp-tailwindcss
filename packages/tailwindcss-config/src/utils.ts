export {
  defu,
  defuOverrideArray,
} from '@weapp-tailwindcss/shared'
