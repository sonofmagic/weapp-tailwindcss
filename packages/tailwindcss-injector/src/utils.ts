export {
  defu,
  defuOverrideArray,
  isRegexp,
  regExpTest,
} from '@weapp-tailwindcss/shared'

export {
  md5,
  removeAllExtensions as removeFileExtension,
} from '@weapp-tailwindcss/shared/node'
