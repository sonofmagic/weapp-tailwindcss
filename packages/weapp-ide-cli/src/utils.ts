import process from 'node:process'
import path from 'pathe'
import { execa } from 'execa'
import type { AliasEntry } from './types'

export async function execute(cliPath: string, argv: string[]) {
  const task = execa(cliPath, argv)

  task?.stdout?.pipe(process.stdout)

  await task
  // 调用返回码为 0 时代表正常，为 -1 时错误。
  // task.exitCode === 0
  // task.exitCode === -1
}

export function resolvePath(filePath: string) {
  if (path.isAbsolute(filePath)) {
    return filePath
  }
  else {
    return path.resolve(process.cwd(), filePath)
  }
}

function alias(argv: string[], entry: AliasEntry) {
  let findIdx = argv.indexOf(entry.find)
  // alias -p as --project
  if (findIdx > -1) {
    argv[findIdx] = entry.replacement
  }
  else {
    findIdx = argv.indexOf(entry.replacement)
  }

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

function pathCompat(argv: string[], option: string) {
  const findIdx = argv.indexOf(option)

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

export function createAlias(entry: AliasEntry) {
  return function (argv: string[]) {
    return alias(argv, entry)
  }
}

export function createPathCompat(option: string) {
  return function (argv: string[]) {
    return pathCompat(argv, option)
  }
}
