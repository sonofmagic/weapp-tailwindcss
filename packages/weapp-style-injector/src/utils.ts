import type { PerFileImportResolver } from './core'

export function toArray<T>(value: T | T[] | null | undefined): Array<NonNullable<T>> {
  if (value == null) {
    return []
  }
  return (Array.isArray(value) ? value : [value]) as Array<NonNullable<T>>
}

export function mergePerFileResolvers(
  resolvers: Array<PerFileImportResolver | null | undefined>,
): PerFileImportResolver | undefined {
  const activeResolvers = resolvers.filter((resolver): resolver is PerFileImportResolver => typeof resolver === 'function')

  if (activeResolvers.length === 0) {
    return undefined
  }

  return (fileName: string) => activeResolvers.flatMap(resolver => toArray(resolver(fileName)))
}

export function ensurePosix(value: string): string {
  return value.replace(/\\/g, '/')
}

export function normalizeRoot(root: string): string {
  const trimmed = root.trim().replace(/^[./\\]+/, '').replace(/\\+/g, '/')
  return trimmed.endsWith('/') ? trimmed.slice(0, -1) : trimmed
}

export function normalizeRelativeImport(target: string): string {
  if (target.startsWith('.') || target.startsWith('/')) {
    return target
  }
  return `./${target}`
}
