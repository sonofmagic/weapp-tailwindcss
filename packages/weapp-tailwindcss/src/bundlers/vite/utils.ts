import type { ExistingRawSourceMap } from 'rollup'
import path from 'node:path'
import process from 'node:process'

export function slash(p: string): string {
  return p.replace(/\\/g, '/')
}

export const isWindows = process.platform === 'win32'

const cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`
export const cssLangRE = new RegExp(cssLangs)
export function isCSSRequest(request: string): boolean {
  return cssLangRE.test(request)
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

const postfixRE = /[?#].*$/
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '')
}

export async function formatPostcssSourceMap(
  rawMap: ExistingRawSourceMap,
  file: string,
): Promise<ExistingRawSourceMap> {
  const inputFileDir = path.dirname(file)

  const sources = rawMap.sources.map((source) => {
    const cleanSource = cleanUrl(decodeURIComponent(source))

    // postcss virtual files
    if (cleanSource[0] === '<' && cleanSource.endsWith('>')) {
      return `\0${cleanSource}`
    }

    return normalizePath(path.resolve(inputFileDir, cleanSource))
  })

  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version,
  }
}
