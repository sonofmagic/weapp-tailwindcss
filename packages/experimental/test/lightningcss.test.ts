import { Buffer } from 'node:buffer'
import { transform } from 'lightningcss'
import { getFixture } from './utils'

// https://lightningcss.dev/transpilation.html
// https://github.com/vitejs/vite/blob/b135918b91e8381c50bd2d076d40e9a65fe68bfe/packages/vite/src/node/plugins/css.ts#L2163
// https://github.com/vitejs/vite/blob/b135918b91e8381c50bd2d076d40e9a65fe68bfe/packages/plugin-legacy/src/index.ts#L227
// https://github.com/vitejs/vite/blob/b135918b91e8381c50bd2d076d40e9a65fe68bfe/packages/vite/src/node/plugins/css.ts#L2122
describe('lightningcss', () => {
  it('base', async () => {
    const { code } = await transform({
      code: Buffer.from(await getFixture('base.css')),
      filename: 'style.css',
      targets: {
        chrome: 61,
      },
    })
    expect(code.toString()).toMatchSnapshot()
  })

  it('components', async () => {
    const { code } = await transform({
      code: Buffer.from(await getFixture('components.css')),
      filename: 'style.css',
      targets: {
        chrome: 61,
      },
    })
    expect(code.toString()).toMatchSnapshot()
  })
})
