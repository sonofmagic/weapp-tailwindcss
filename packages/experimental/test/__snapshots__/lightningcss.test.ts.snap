// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lightningcss > base 1`] = `
"@layer base {
  h1 {
    font-size: var(--text-2xl);
  }

  h2 {
    font-size: var(--text-xl);
  }
}
"
`;

exports[`lightningcss > components 1`] = `
"@layer components {
  .card {
    background-color: var(--color-white);
    -webkit-border-radius: var(--rounded-lg);
    border-radius: var(--rounded-lg);
    padding: var(--spacing-6);
    box-shadow: var(--shadow-xl);
  }
}
"
`;
